### PR TITLE
Add Pokemon Let's Go!

### PR DESCRIPTION
**Pull request by KrisXV**

### CODE REVIEW

**Comment by Zarel** [comment]
This pull request **introduces 1 alert** when merging b93c5182c074f3fce9e445a0678f37729d4d7560 into a03707e9c027676a513d18c73009abda47d7e1a2 - [view on LGTM.com](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown/rev/pr-5af9e58b1fa0af507aa219ac3818940a8659d670)

**new alerts:**

* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by TheImmortal** [comment]
Make sure none of this is legal is Hackmons, or any OM for that matter, as long as standard tiers follow USUM the so do non-standard tiers. Let’s Go is a unique format.

If it’s possible to put more data in the mod rather than the main data files, that would help.

**Comment by xfix** [comment]
Considering we have stuff like Pokestar Studio mons, CAP, Missingno. and some other things in data files, I don't see a problem with Let's Go stuff in data files, as long they are properly marked as `nonStandard`.

**Comment by KrisXV** [comment]
I was planning on doing just that as soon as I got computer access

File: data/formats-data.js [hunk 1]
```diff
@@ -442,6 +442,15 @@ let BattleFormatsData = {
...
 		doublesTier: "DUU",
 	},
+	pikachustarter: {
+		eventPokemon: [
+			{"generation": 7, "level": 5, "perfectIVs": 6, "moves": ["thundershock", "tailwhip", "growl"], "pokeball": "pokeball"},
+		],
+		eventOnly: true,
+		isNonstandard: true,
+		isUnreleased: true,
+		tier: "Unreleased",
```
Comment (xfix) [inline data/formats-data.js]: It's Invalid in context of Gen7, not Unreleased

Comment (Zarel) [reply data/formats-data.js]: `Illegal` is I think the text we use?

Comment (KrisXV) [reply data/formats-data.js]: The tier doesn’t matter, it’s the `isNonstandard` tag. Not to mention meltan and melmetal are confirmed unreleased.

**Comment by TheImmortal** [comment]
Could there be a way to turn AVs on and off? Such as an AllowAVs rule.

File: data/formats-data.js (hunk 1 above)
Comment (KrisXV) [reply data/formats-data.js]: That wasn’t what I meant.. but alright. I don’t feel comfortable with tiering these as illegal.

**Comment by KrisXV** [comment]
It’s possible but it requires doing the opposite of what was originally requested ie making as little code outside of mods files as possible. I was going to initially do that but after discussing with Marty I decided to wait on it a bit to see how the one test format works.

**Comment by KrisXV** [comment]
@theimmortal

File: data/formats-data.js (hunk 1 above)
Comment (Zarel) [reply data/formats-data.js]: They're illegal in USUM. This is `data/format-data.js`, which is a USUM file. You can tier them `Unreleased` in `mods/lgpe/format-data.js`, but not in this one.

File: mods/letsgo/moves.js [hunk 2]
```diff
@@ -52,9 +52,9 @@ let BattleMovedex = {
 		onHit: function (target, source, effect) {
 			let moves = [];
 			for (let i in exports.BattleMovedex) {
-				let move = exports.BattleMovedex[i];
+				let move = this.getMove(i);
```
Comment (Zarel) [inline mods/letsgo/moves.js]: You can manually cast, and I'd recommend it.

Comment (Zarel) [reply mods/letsgo/moves.js]:
```suggestion
				let move = /** @type {MoveData} */ (exports.BattleMovedex[i]);
```

**Comment by KrisXV** [comment]
:/

**Comment by Zarel** [comment]
"Property `gen` doesn't exist on type `MoveData`" apparently. It should probably be added, considering it literally is there.

**Comment by Zarel** [comment]
Oh, apparently it isn't, and is actually auto-inferred from `.num`. So TypeScript did catch an actual bug there.

**Comment by KrisXV** [comment]
for now im going to leave it as this so the build passes

**Comment by Zarel** [comment]
No you should definitely leave it; I assumed from the "TypeScript why" commit description that TypeScript was being overzealous, but no it did actually catch a real bug.

File: sim/pokemon.js [hunk 3]
```diff
@@ -277,6 +277,11 @@ class Pokemon {
 		this.boosts = {atk: 0, def: 0, spa: 0, spd: 0, spe: 0, accuracy: 0, evasion: 0};
 		/**@type {{[k: string]: number}} */
 		this.stats = {atk: 0, def: 0, spa: 0, spd: 0, spe: 0};
+		/**
+		 * Commented out until a use is found for Combat Power in Let's Go
+		 * @type {number?}
+		 * this.combatPower = null;
```
Comment (Zarel) [inline sim/pokemon.js]: This definitely doesn't need to be here at all. Just recalculate Combat Power every time it's needed.

File: mods/letsgo/rulesets.js [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+				}
+			}
+
+			// ----------- legality line ------------------------------------------
+			if (!this.getRuleTable(format).has('-illegal')) return problems;
+			// everything after this line only happens if we're doing legality enforcement
+
+			// Pokemon cannot have more than 1200 total Awakening Values
+			if (totalAV > 1200) {
+				problems.push((set.name || set.species) + " has more than 1200 total Awakening Values.");
```
Comment (Zarel) [inline mods/letsgo/rulesets.js]: ...why not count them as 200 each instead of 1200 total?

Comment (KrisXV) [reply mods/letsgo/rulesets.js]: Each stat is already clamped from 0 to 200, this is just to check for any weird situations that might possibly happen. I guess I can do that though

Comment (Zarel) [reply mods/letsgo/rulesets.js]: I wouldn't. I also wouldn't clamp it, but give an error message if it's over 200.

**Pull request merged**